### PR TITLE
Lacy Draper - Scissors

### PR DESCRIPTION
**Pull request by LacyDraper**

### CODE REVIEW

File: linked_list/linked_list.py [hunk 1]
```diff
@@ -10,70 +10,145 @@ def __init__(self, value, next_node = None):
...
       self.head = None # keep the head private. Not accessible outside this class
-
+    
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first(self):
```
Comment (CheezItMan) [inline linked_list/linked_list.py:18]: 👍

File: linked_list/linked_list.py [hunk 2]
```diff
@@ -10,70 +10,145 @@ def __init__(self, value, next_node = None):
...
+
 
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
```
Comment (CheezItMan) [inline linked_list/linked_list.py:30]: 👍

File: linked_list/linked_list.py [hunk 3]
```diff
@@ -10,70 +10,145 @@ def __init__(self, value, next_node = None):
...
+        self.head = new_node
+        
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
```
Comment (CheezItMan) [inline linked_list/linked_list.py:40]: 👍

File: linked_list/linked_list.py [hunk 4]
```diff
@@ -10,70 +10,145 @@ def __init__(self, value, next_node = None):
...
+            else:
+                current = current.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
```
Comment (CheezItMan) [inline linked_list/linked_list.py:54]: 👍

File: linked_list/linked_list.py [hunk 5]
```diff
@@ -10,70 +10,145 @@ def __init__(self, value, next_node = None):
...
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        current = self.head
+        if current == None:
+            return None
+        counter = 0
+        while counter != index:
+            current = current.next
+            counter += 1
```
Comment (CheezItMan) [inline linked_list/linked_list.py:75]: You should also have a test in here to check to see if current is None.  If it is you can return None.

It's possible that `index` is > the length of the list.  Something the tests didn't catch.

File: linked_list/linked_list.py [hunk 6]
```diff
@@ -10,70 +10,145 @@ def __init__(self, value, next_node = None):
...
+            
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(n)
+    
     def get_last(self):
```
Comment (CheezItMan) [inline linked_list/linked_list.py:84]: 👍

File: linked_list/linked_list.py [hunk 7]
```diff
@@ -10,70 +10,145 @@ def __init__(self, value, next_node = None):
...
+            current = current.next
+        return current.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: 0(n) because it has to move through each existing node
+    # Space Complexity: O(n)
+    
     def add_last(self, value):
```
Comment (CheezItMan) [inline linked_list/linked_list.py:96]: 👍 the space complexity is O(1) since you are not creating a new list or data structure in the function itself.

File: linked_list/linked_list.py [hunk 8]
```diff
@@ -10,70 +10,145 @@ def __init__(self, value, next_node = None):
...
+            return
+        current_node = self.head
+        while current_node.next:
+           current_node = current_node.next
+        current_node.next = new_node
+            
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
```
Comment (CheezItMan) [inline linked_list/linked_list.py:109]: 👍

File: linked_list/linked_list.py [hunk 9]
```diff
@@ -10,70 +10,145 @@ def __init__(self, value, next_node = None):
...
+        if current_node.next.value == value:
+            current_node = self.head
+        
+        while current_node.next:
+            if current_node.next.value == value:
+                if not current_node.next.next:
+                    current_node.next = None
+                else:
+                    current_node = current_node.next.next
+                    return
```
Comment (CheezItMan) [inline linked_list/linked_list.py:138]: The inner `if` is not needed

```suggestion
            if current_node.next.value == value:
                current_node = current_node.next.next
                return
```

File: linked_list/linked_list.py [hunk 10]
```diff
@@ -10,70 +10,145 @@ def __init__(self, value, next_node = None):
...
+
+
+
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def visit(self):
```
Comment (CheezItMan) [inline linked_list/linked_list.py:152]: 👍

File: linked_list/linked_list.py [hunk 11]
```diff
@@ -86,10 +161,22 @@ def visit(self):
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def reverse(self):
```
Comment (CheezItMan) [inline linked_list/linked_list.py:166]: 👍 Nicely done, but the space complexity is O(1) because no matter how big the list is, you're only making 3 variables, none of which is a larger data structure.

**Review by CheezItMan — COMMENTED** [review]
Well done Lacy, I had a few comments, but they were minor.  This is great work.
